### ISSUE TITLE
NoHostAvailableException during cassandraClient.connect leads to failure in Future.fail()

### ISSUE THREAD

**Issue by triplem** [vert-x3/vertx-cassandra-client#21]
```
        try {
            client = client.connect(CASSANDRA_KEYSPACE) { ar ->
                if (ar.failed()) {
                    logger.error("ALL IS BAD....")
                    startFuture.fail(ar.cause())
                } else {
                    repository = Repository(client)

                    vertx.eventBus().consumer<JsonObject>(ADDRESS, this::handle)
                    startFuture.complete()
                }
            }
        } catch (e: Exception) {
            logger.warn("catched exception")
            startFuture.fail(e)
        }
```

The above code leads to an exception, if no cassandra server is available (NoHostAvailableException). This exception is catched in the given catch block. If startFuture.complete() is called, all is fine (well, then the verticle is deployed, even though it shouldn't), but if startFuture.fail(e) is called, the event is not fetched in my MainVerticle.

**Comment by triplem** [comment]
Thanks for the fix, unfortunately this is still not working properly. The exception is catched and the future is failed, perfect. But in my MainVertical, I still do not get this state.

```
            vertx.deployVerticle("ServiceVerticle") { ar ->
                if (ar.failed()) {
                    logger.error("Could not start serviceVerticle", ar.cause())
                } else {
                    logger.error("........................................")
                    logger.error("ar: ", ar.result())
                }
            }
```
In the above code, on an failed future, nothing really happens.

**Comment by Sammers21** [comment]
I not sure if I understood you correctly. Can you provide a reproducer in order to make it obvious?

**Comment by Sammers21** [comment]
Is this something, which can be fixed in Cassandra client? The future is failed by Cassandra client, so this is what is expected. If I got you right, the situation is similar to this:
```java
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.Vertx;

public class Main {

  public static void main(String[] args) {
    Vertx vertx = Vertx.vertx();
    vertx.deployVerticle(new A());
  }

  static class A extends AbstractVerticle {
    @Override
    public void start() {
      vertx.deployVerticle(new B(), deployResult -> {
        if (deployResult.succeeded()) {
          System.out.println("deployed");
        } else {
          System.out.println("not deployed, reason: " + deployResult.cause());
        }
      });
    }
  }

  static class B extends AbstractVerticle {
    @Override
    public void start(Future<Void> startFuture) {
      startFuture.fail("Something is wrong");
    }
  }
}
```
And `not deployed, reason: ....` is printed in this case.

**Comment by triplem** [comment]
I have provided a "reproducer" in https://github.com/triplem/vertx-cassandra-client/tree/3.6 (take a look into the examples). But from your example above, this is exactly what is "NOT" happening.

**Comment by Sammers21** [comment]
Just created an issue for this: https://github.com/eclipse-vertx/vert.x/issues/2809

**Comment by Sammers21** [comment]
it seems that the issue will be solved in the #16 PR
